infrastructure/src/main/java/org/corfudb/infrastructure/RequestHandlerMethods.java

```diff
@@ -0,0 +1,146 @@
+package org.corfudb.infrastructure;
+
+import com.codahale.metrics.Timer;
+import io.netty.channel.ChannelHandlerContext;
+import java.lang.invoke.LambdaMetafactory;
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.util.Arrays;
+import java.util.EnumMap;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import javax.annotation.Nonnull;
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.runtime.protocol.proto.CorfuProtocol.Request;
+import org.corfudb.runtime.protocol.proto.CorfuProtocol.MessageType;
+import org.corfudb.runtime.exceptions.unrecoverable.UnrecoverableCorfuError;
+import org.corfudb.util.CorfuComponent;
+import org.corfudb.util.MetricsUtils;
+
+@Slf4j
+public class RequestHandlerMethods {
+
+    private final Map<MessageType, String> timerNameCache = new HashMap<>();
+
+    /** The handler map. */
+    private final Map<MessageType, HandlerMethod> handlerMap;
+
+    /**
+     * A functional interface for server request handlers. Server request handlers should
+     * be fast and not block. If a handler blocks for an extended period of time, it will
+     * exhaust the server's thread pool. I/O and other long operations should be handled
+     * on another thread.
+     */
+    @FunctionalInterface
+    public interface HandlerMethod {
+        void handle(@Nonnull Request req,
+                    @Nonnull ChannelHandlerContext ctx,
+                    @Nonnull IRequestRouter r);
+    }
+
+    /** Get the types of requests this handler will handle.
+     *
+     * @return  A set containing the types of requests this handler will handle.
+     */
+    public Set<MessageType> getHandledTypes() {
+        return handlerMap.keySet();
+    }
+
+    /** Construct a new instance of RequestHandlerMethods. */
+    public RequestHandlerMethods() {
+        handlerMap = new EnumMap<>(MessageType.class);
+    }
+
+    /** Handle an incoming Corfu request message.
+     *
+     * @param req       The request message to handle.
+     * @param ctx       The channel handler context.
+     * @param r         The server router.
+     */
+    @SuppressWarnings("unchecked")
+    public void handle(Request req, ChannelHandlerContext ctx, IRequestRouter r) {
+        final HandlerMethod handler = handlerMap.get(req.getHeader().getType());
```

Comment:
Having the type in the header seems redundant. Would it be possible to bootstrap the handlers this way?


```
public @interface RequestHandler {
    /**
     * Returns the message type
     * @return the type of Corfu message request
     */
    CorfuMessage.RequestMsg.PayloadCase type();
}

@RequestHandler(type = CorfuMessage.RequestMsg.PayloadCase.AUTHENTICATE_REQUEST)

final HandlerMethod handler = handlerMap.get(req.getPayloadCase());